src/compile.c

```diff
@@ -0,0 +1,403 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "c2mir/c2mir.h"
+#include "cache.h"
+#include "compile.h"
+#include "decode.h"
+#include "mir-gen.h"
+#include "mir.h"
+#include "riscv_private.h"
+#include "utils.h"
+
+typedef struct {
+    char *code;
+    size_t code_size;
+    size_t curr;
+} jit_item_t;
+
+typedef struct {
+    MIR_context_t ctx;
+    struct c2mir_options *options;
+    uint8_t debug_level;
+    uint8_t optimize_level;
+} riscv_jit_t;
+
+typedef struct {
+    char *name;
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define GOLDEN_RATIO_32 0x61C88647
```

Comment:
Both `GOLDEN_RATIO_32` and `HASH` already exist in `src/cache.c`. We shall put them to somewhere such as `src/utils.h`.